File: java/spring-boot/src/main/java/com/backendtemplate/services/ReservationService.java
package com.backendtemplate.services;

import com.backendtemplate.contracts.Reservation;
import com.backendtemplate.contracts.ReservationResult;
import com.backendtemplate.repositories.ReservationRepository;
import java.time.Clock;
import java.util.UUID;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Profile("!no-db")
public class ReservationService {
    private final ReservationRepository repository;
    private final Clock clock;

    public ReservationService(ReservationRepository repository, Clock clock) {
        this.repository = repository;
        this.clock = clock;
    }

    @Transactional(rollbackFor = Exception.class)
    public ReservationResult reserve(String productId, String key) {
        var replay = repository.findMatchingReplay(key, productId);
        if (replay.isPresent()) {
            return new ReservationResult(replay.get(), true);
        }
        repository.claim(key);
        repository.decreaseStock(productId);
        var reservation = new Reservation(UUID.randomUUID().toString().replace("-", ""), productId, clock.instant());
        repository.saveReservationAndReplay(reservation, key);
        return new ReservationResult(reservation, false);
    }

    @Transactional(readOnly = true, rollbackFor = Exception.class)
    public ReservationResult replay(String productId, String key) {
        return new ReservationResult(repository.findMatchingReplay(key, productId).orElseThrow(), true);
    }
}

File: kotlin/spring-boot/src/main/kotlin/com/backendtemplate/services/ReservationService.kt
package com.backendtemplate.services

import com.backendtemplate.contracts.Reservation
import com.backendtemplate.contracts.ReservationResult
import com.backendtemplate.exceptions.ReservationFailure
import com.backendtemplate.exceptions.ReservationFailure.Reason
import com.backendtemplate.repositories.ReservationRepository
import com.backendtemplate.validation.validateProductExists
import com.backendtemplate.validation.validateReplayProduct
import java.time.Clock
import java.util.UUID
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Profile("!no-db")
class ReservationService(
    private val repository: ReservationRepository,
    private val clock: Clock,
) {
    @Transactional(rollbackFor = [Exception::class])
    fun reserve(productId: String, key: String): ReservationResult {
        val replay = repository.findReplay(key)
        if (replay != null) {
            validateReplayProduct(productId, replay.productId)
            return ReservationResult(reservation = replay, replayed = true)
        }
        repository.claim(key)
        if (!repository.decreaseStockIfAvailable(productId)) {
            validateProductExists(repository.productExists(productId))
            throw ReservationFailure(Reason.SOLD_OUT)
        }
        val reservation = Reservation(
            reservationId = UUID.randomUUID().toString().replace("-", ""),
            productId = productId,
            createdAt = clock.instant(),
        )
        repository.saveReservationAndReplay(reservation, key)
        return ReservationResult(reservation = reservation, replayed = false)
    }

    @Transactional(readOnly = true, rollbackFor = [Exception::class])
    fun replay(productId: String, key: String): ReservationResult {
        val replay = repository.findReplay(key) ?: error("Committed replay is missing")
        validateReplayProduct(productId, replay.productId)
        return ReservationResult(reservation = replay, replayed = true)
    }
}
